### ISSUE TITLE
Problem when multicore = T

### ISSUE THREAD

**Issue by HanZhang-psych** [simsem/simsem#64]
Hello!

First of all, this is a great project! I ran into a problem when playing around with the example code:

```
popNull <- "
f1 =~ 0.7*y1 + 0.7*y2 + 0.7*y3 + 0.7*y4 + 0.7*y5 + 0.7*y6
f1 ~~ 1*f1
y1 ~~ 0.51*y1
y2 ~~ 0.51*y2
y3 ~~ 0.51*y3
y4 ~~ 0.51*y4
y5 ~~ 0.51*y5
y6 ~~ 0.51*y6
"

analyzeNull <- "
f1 =~ y1 + y2 + y3 + y4 + y5 + y6
"

Output.NULL <- sim(100, analyzeNull, n=200, generate=popNull, std.lv=TRUE, lavaanfun = "cfa", multicore = T)

summary(Output.NULL)
```

This would return:
```
RESULT OBJECT
Model Type
[1] "lavaan"
Error in clean(object) : 
  All replications in the result object are not convergent. Thus, the result object cannot be used.
```

Using ```multicore = F``` seems fine. The output shows that all replications are successfully converged.

Any idea why this happened?

Thanks!

```
sessioninfo::session_info()
- Session info ---------------------------------------------------------------------------------------------------------------------------------------------------------
 setting  value                       
 version  R version 4.0.0 (2020-04-24)
 os       Windows 10 x64              
 system   x86_64, mingw32             
 ui       RStudio                     
 language (EN)                        
 collate  English_United States.1252  
 ctype    English_United States.1252  
 tz       America/New_York            
 date     2020-07-19                  

- Packages -------------------------------------------------------------------------------------------------------------------------------------------------------------
 package     * version  date       lib source        
 assertthat    0.2.1    2019-03-21 [1] CRAN (R 4.0.0)
 cli           2.0.2    2020-02-28 [1] CRAN (R 4.0.0)
 crayon        1.3.4    2017-09-16 [1] CRAN (R 4.0.0)
 fansi         0.4.1    2020-01-08 [1] CRAN (R 4.0.0)
 glue          1.4.1    2020-05-13 [1] CRAN (R 4.0.0)
 lavaan      * 0.6-6    2020-05-13 [1] CRAN (R 4.0.2)
 MASS          7.3-51.5 2019-12-20 [2] CRAN (R 4.0.0)
 mnormt        2.0.0    2020-06-02 [1] CRAN (R 4.0.0)
 pbivnorm      0.6.0    2015-01-23 [1] CRAN (R 4.0.0)
 rstudioapi    0.11     2020-02-07 [1] CRAN (R 4.0.0)
 sessioninfo   1.1.1    2018-11-05 [1] CRAN (R 4.0.0)
 simsem      * 0.5-15   2020-02-12 [1] CRAN (R 4.0.2)
 snow          0.4-3    2018-09-14 [1] CRAN (R 4.0.0)
 tmvnsim       1.0-2    2016-12-15 [1] CRAN (R 4.0.0)
 withr         2.2.0    2020-04-20 [1] CRAN (R 4.0.0)
```

**Comment by Suppanut** [comment]
To fix this, you may want to add `parallel::clusterExport(cl, lavaanfun)` under `cl <- parallel::makeCluster(rep("localhost", numProc),  type = "SOCK")`

To illustrate:

```
library(simsem)
popNull <- "
f1 =~ 0.7*y1 + 0.7*y2 + 0.7*y3 + 0.7*y4 + 0.7*y5 + 0.7*y6
f1 ~~ 1*f1
y1 ~~ 0.51*y1
y2 ~~ 0.51*y2
y3 ~~ 0.51*y3
y4 ~~ 0.51*y4
y5 ~~ 0.51*y5
y6 ~~ 0.51*y6
"

analyzeNull <- "
f1 =~ y1 + y2 + y3 + y4 + y5 + y6
"

simConds <- list(list(NULL,200,NULL,NULL,NULL,FALSE),list(NULL,200,NULL,NULL,NULL,FALSE))
lavaanfun <- "cfa"

## single core 
lapply(simConds, simsem:::runRep, model = list(model = analyzeNull, std.lv = TRUE), generateO = list(model = popNull), lavaanfun = lavaanfun)

## multicores
numProc <- parallel::detectCores()-1
cl <- parallel::makeCluster(rep("localhost", numProc),  type = "SOCK")
### this will not work
parallel::clusterApplyLB(cl, simConds, simsem:::runRep, model = list(model = analyzeNull, std.lv = TRUE), generateO = list(model = popNull), lavaanfun = lavaanfun)

## Export lavaanfun
parallel::clusterExport(cl, lavaanfun)
### this will work
parallel::clusterApplyLB(cl, simConds, simsem:::runRep, model = list(model = analyzeNull, std.lv = TRUE), generateO = list(model = popNull), lavaanfun = lavaanfun)
```

**Comment by TDJorgensen** [comment]
Thank you Suppanut!  I expected a call to `parallel::clusterExport()` would be involved in the solution, but I couldn't guess which among many objects might need to be passed.  I still expect I might need to pass the remaining `runRep()` arguments, but for now, I added this update to the development version (0.5-15.904), which can be installed from GitHub via:

```
devtools::install_github("simsem/simsem/simsem")
```

The help-page examples work on Windows with `multicore=TRUE`, so I expect this resolves the OP's example as well.

**Comment by AJThurston** [comment]
I was experiencing the same issue and used @Suppanut's fix and it worked perfectly for me!  Thanks so much!

**Comment by iwoasned** [comment]
Hi everyone,

I am also running into problems when using `multicore = T`. Specifically, simsem won't use my costum output function when using parallel processing.

For example:
```
library(simsem)

outfun <- function(output){
  
  par <- parameterestimates(output, standardized = TRUE)
  
  std.lambda.g <- par[par$lhs == 'g' & par$op == '=~', 'std.all']
  std.lambda.spec <- par[par$lhs != 'g' & par$op == '=~', 'std.all']
  est.ecv <- (sum(std.lambda.g^2))/(sum(std.lambda.g^2) + sum(std.lambda.spec^2))
  
  return(est.ecv)
  
}


lambda <- matrix(0, nrow = 9, ncol = 4)
lambda[, 1] <- 0.7
lambda[1:3, 2] <- 0.4
lambda[4:6, 3] <- 0.4
lambda[7:9, 4] <- 0.4

lambda.var <- lambda
lambda.var[lambda != 0] <- NA
LY <- bind(lambda.var, lambda)

phi <- diag(nrow = 4, ncol = 4)
RPS <- binds(phi, 0)

theta <- diag(0,9,9)
RTE <- binds(theta)

CFA.model <- model(LY = LY, RPS = RPS, RTE = RTE, modelType = "CFA", facLab = c('g', 's1', 's2', 's3'))

## this works
simres <- sim(model = CFA.model, n = 300, nRep = 10, outfun = outfun)
getExtraOutput(simres)

## this won't work
simres <- sim(model = CFA.model, n = 300, nRep = 10, outfun = outfun, multicore = TRUE, numProc = 4)
getExtraOutput(simres)

```

I tried Suppanut's fix, but I couldn't make it work. I am using simsem 0.5-16.906.

Thanks for your help.

**Comment by TDJorgensen** [comment]
> simsem won't use my costum output function when using parallel processing

That is one of the things I meant when I said 

> I still expect I might need to pass the remaining `runRep()` arguments

Unfortunately, I have tried to pass that using `clusterExport()`, but it still did not work on Windows.

For these parallel problems on Windows, I am thinking of utilizing the `lavaanList()` for its much more reliable use of `parallel` options (and assuming of course that `lavaan` is the analysis engine).  But that would be at an undetermined time in the future.

**Comment by Suppanut** [comment]
Hi @iwoasned, in your function, try `lavaan::parameterestimates` instead of `parameterestimates`.

**Comment by iwoasned** [comment]
Thank you very much @Suppanut!

**Comment by TDJorgensen** [comment]
> try `lavaan::parameterestimates` 

oh, of course. I was so hung up on `clusterExport()`, I forgot to look in the `outfun()` definition itself to make sure everything would be accessible on each node.  Thank you @Suppanut!

**Comment by isaactpetersen** [comment]
I get an error when using `multicore = TRUE` on Windows.

```
Error in checkForRemoteErrors(val) : 
  one node produced an error: no method found for function 'coef' and signature lavaan.mi
```

See `reprex()` below:
``` r
library("lavaan")
#> This is lavaan 0.6-11
#> lavaan is FREE software! Please report any bugs.
library("simsem")
#> 
#> #################################################################
#> This is simsem 0.5-16
#> simsem is BETA software! Please report any bugs.
#> simsem was first developed at the University of Kansas Center for
#> Research Methods and Data Analysis, under NSF Grant 1053160.
#> #################################################################
#> 
#> Attaching package: 'simsem'
#> The following object is masked from 'package:lavaan':
#> 
#>     inspect
library("semTools")
#> 
#> ###############################################################################
#> This is semTools 0.5-5
#> All users of R (or SEM) are invited to submit functions or ideas for functions.
#> ###############################################################################
library("mice")
#> 
#> Attaching package: 'mice'
#> The following object is masked from 'package:stats':
#> 
#>     filter
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind

# Specify analysis model
analysisModelSyntax <- '
 # Measurement model factor loadings
 ind60 =~ NA*x1 + x2 + x3
 dem60 =~ NA*y1 + y2 + y3 + y4
 dem65 =~ NA*y5 + y6 + y7 + y8

 # Regression paths
 dem60 ~ ind60
 dem65 ~ ind60 + dem60
 
 # Fix latent means to zero
 ind60 ~ 0
 dem60 ~ 0
 dem65 ~ 0
 
 # Fix latent variances to one
 ind60 ~~ 1*ind60
 dem60 ~~ 1*dem60
 dem65 ~~ 1*dem65
 
 # Estimate residual variances of manifest variables
 x1 ~~ x1
 x2 ~~ x2
 x3 ~~ x3
 y1 ~~ y1
 y2 ~~ y2
 y3 ~~ y3
 y4 ~~ y4
 y5 ~~ y5
 y6 ~~ y6
 y7 ~~ y7
 y8 ~~ y8
 
 # Free intercepts of manifest variables
 x1 ~ intx1*1
 x2 ~ intx2*1
 x3 ~ intx3*1
 y1 ~ inty1*1
 y2 ~ inty2*1
 y3 ~ inty3*1
 y4 ~ inty4*1
 y5 ~ inty5*1
 y6 ~ inty6*1
 y7 ~ inty7*1
 y8 ~ inty8*1
'

# Specify population model

populationModel <- '
 # Specify measurement model factor loadings
 ind60 =~ .75*x1 + .75*x2 + .75*x3
 dem60 =~ .75*y1 + .75*y2 + .75*y3 + .75*y4
 dem65 =~ .75*y5 + .75*y6 + .75*y7 + .75*y8

 # Specify regression coefficients
 dem60 ~ .4*ind60
 dem65 ~ .2*ind60 + .45*dem60
 
 # Fix latent means to zero
 ind60 ~ 0
 dem60 ~ 0
 dem65 ~ 0
 
 # Fix latent variances to one
 ind60 ~~ 1*ind60
 dem60 ~~ 1*dem60
 dem65 ~~ 1*dem65
 
 # Specify residual variances of manifest variables
 x1 ~~ (1-.75^2)*x1
 x2 ~~ (1-.75^2)*x2
 x3 ~~ (1-.75^2)*x3
 y1 ~~ (1-.75^2)*y1
 y2 ~~ (1-.75^2)*y2
 y3 ~~ (1-.75^2)*y3
 y4 ~~ (1-.75^2)*y4
 y5 ~~ (1-.75^2)*y5
 y6 ~~ (1-.75^2)*y6
 y7 ~~ (1-.75^2)*y7
 y8 ~~ (1-.75^2)*y8
 
 # Specify intercepts of manifest variables
 x1 ~ 0*1
 x2 ~ 0*1
 x3 ~ 0*1
 y1 ~ 0*1
 y2 ~ 0*1
 y3 ~ 0*1
 y4 ~ 0*1
 y5 ~ 0*1
 y6 ~ 0*1
 y7 ~ 0*1
 y8 ~ 0*1
'

# Specify missingness
percentMissingByVariable <- '
  x1 ~ p(0.10)
  x2 ~ p(0.10)
  x3 ~ p(0.10)
  y1 ~ p(0.10)
  y2 ~ p(0.10)
  y3 ~ p(0.10)
  y4 ~ p(0.10)
  y5 ~ p(0.10)
  y6 ~ p(0.10)
  y7 ~ p(0.10)
  y8 ~ p(0.10)
'

# Multiple Imputation
missingnessModelMI <- miss(logit = percentMissingByVariable, m = 5)

# This works
output <- simsem::sim(n = 150,
                      model = analysisModelSyntax,
                      generate = populationModel,
                      miss = missingnessModelMI)
#> Progress: 1 / 1 
#> 
#> The RIV will exceed 1 whenever between-imputation variance exceeds within-imputation variance (when FMI(1) > 50%).
#> 
#> lavaan.mi object based on 5 imputed data sets. 
#> See class?lavaan.mi help page for available methods. 
#> 
#> Convergence information:
#> The model converged on 5 imputed data sets 
#> 
#> Rubin's (1987) rules were used to pool point and SE estimates across 5 imputed data sets, and to calculate degrees of freedom for each parameter's t test and CI.

# This doesn't work
output <- simsem::sim(n = 150,
                      model = analysisModelSyntax,
                      generate = populationModel,
                      miss = missingnessModelMI,
                      multicore = TRUE)
#> Progress tracker is not available when 'multicore' is TRUE.
#> Error in checkForRemoteErrors(val): one node produced an error: no method found for function 'coef' and signature lavaan.mi
```
<sup>Created on 2022-04-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Here is my `sessionInfo()`:
```
sessionInfo()
#> R version 4.2.0 (2022-04-22 ucrt)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 19044)
#> 
#> Matrix products: default
#> 
#> Random number generation:
#>  RNG:     L'Ecuyer-CMRG 
#>  Normal:  Inversion 
#>  Sample:  Rejection 
#>  
#> locale:
#> [1] LC_COLLATE=English_United States.utf8 
#> [2] LC_CTYPE=English_United States.utf8   
#> [3] LC_MONETARY=English_United States.utf8
#> [4] LC_NUMERIC=C                          
#> [5] LC_TIME=English_United States.utf8    
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] mice_3.14.0    semTools_0.5-5 simsem_0.5-16  lavaan_0.6-11 
#> 
#> loaded via a namespace (and not attached):
#>  [1] Rcpp_1.0.8.3      mvtnorm_1.1-3     lattice_0.20-45   tidyr_1.2.0      
#>  [5] snow_0.4-4        zoo_1.8-10        assertthat_0.2.1  digest_0.6.29    
#>  [9] utf8_1.2.2        R6_2.5.1          backports_1.4.1   reprex_2.0.1     
#> [13] stats4_4.2.0      evaluate_0.15     coda_0.19-4       highr_0.9        
#> [17] pillar_1.7.0      rlang_1.0.2       multcomp_1.4-18   rstudioapi_0.13  
#> [21] R.utils_2.11.0    R.oo_1.24.0       Matrix_1.4-1      pbivnorm_0.6.0   
#> [25] rmarkdown_2.13    styler_1.7.0      splines_4.2.0     stringr_1.4.0    
#> [29] broom_0.8.0       compiler_4.2.0    xfun_0.30         pkgconfig_2.0.3  
#> [33] mnormt_2.0.2      tmvnsim_1.0-2     htmltools_0.5.2   tidyselect_1.1.2 
#> [37] tibble_3.1.6      codetools_0.2-18  fansi_1.0.3       crayon_1.5.1     
#> [41] dplyr_1.0.8       withr_2.5.0       MASS_7.3-57       R.methodsS3_1.8.1
#> [45] grid_4.2.0        xtable_1.8-4      lifecycle_1.0.1   DBI_1.1.2        
#> [49] magrittr_2.0.3    estimability_1.3  cli_3.2.0         stringi_1.7.6    
#> [53] fs_1.5.2          ellipsis_0.3.2    vctrs_0.4.1       generics_0.1.2   
#> [57] sandwich_3.0-1    TH.data_1.1-0     tools_4.2.0       R.cache_0.15.0   
#> [61] glue_1.6.2        purrr_0.3.4       emmeans_1.7.3     parallel_4.2.0   
#> [65] fastmap_1.1.0     survival_3.3-1    yaml_2.3.5        knitr_1.38
```

**Comment by Suppanut** [comment]
I think loading `semTools` will help.  For me, adding `parallel::clusterEvalQ(cl, library("semTools"))` below `parallel::clusterExport(cl, lavaanfun)` in the `simsem::sim()` function solves this issue (or replacing the later with the former).

**Comment by isaactpetersen** [comment]
Thanks @Suppanut. I had loaded `semTools` (see reprex/sessionInfo). Unfortunately, adding that line in the `simsem::sim()` function didn't work for me. I still get the following error:
```
Error in checkForRemoteErrors(val) : 
  50 nodes produced errors; first error: no method found for function 'coef' and signature lavaan.mi
```

**Comment by isaactpetersen** [comment]
That works, thanks @Suppanut! Any way we can incorporate this code change into the package?

**Comment by Suppanut** [comment]
@isaactpetersen One way is to download the package source code, edit it on our computer, then install it via, e.g., `remotes::install_local()`

**Comment by TDJorgensen** [comment]
No need, I just addressed this in the latest commit (https://github.com/simsem/simsem/commit/3e7ef682251facd5b2ec45c29ffd7cc8e720d416), so you can install the development version (0.5-16.909).  

```
devtools::install_github("simsem/simsem/simsem")
```

FYI, without knowing `semTools` is installed on each node, it is safer to use `require()` because `library()` returns an error rather than a warning.

**Comment by isaactpetersen** [comment]
Sounds great, thanks!

**Comment by BorisEckstein** [comment]
Hello everyone

First of all, thank you very much for your engagement!

I am not sure if this is the right place to ask, but my google search did not reveal any better place. I have a similar issue as described above: my analyses won't converge. I tried all recommendations from above that I understood (which unluckily excludes @Suppanut's solution). I basically turned off the multicore argument and installed simsem's GitHub version (0.5-16.909).

I tried to replicate the cases explained in your vignettes 8 (missings) and 12 (non normality). When running them separately, the simulations converge nicely. However, I tried to COMBINE both issues, which resulted in this error message:

Error in clean(object, improper = improper) : 
All replications in the result object are not convergent. Thus, the result object cannot be used.

I tried several things step by step: raising the sample size to n = 5000; raising the iteration number to 5000; change the estimator to "ml".

I think that the simulations SHOULD converge, as both issues are only mildly (2% missing; skewness/kurtosis from -2.5 to +2.5).

**My question**: Is there an issue with my script or could this be a technical problem as discussed in this thread?

Any hints will be appreciated very much!


Here's my script in the latest attempt:

`
library(simsem)
library(lavaan)

# -------------------------------------------------------------------------- #
## Population model ----
# -------------------------------------------------------------------------- #

mod1.pop <- '

# factor loadings fixed @ 1.00 and 0.70 
SP.o.t1 =~ 1.0 * sp.o.t1.1 + .7 * sp.o.t1.2 + .7 * sp.o.t1.3
SP.t.t1 =~ 1.0 * sp.t.t1.1 + .7 * sp.t.t1.2 + .7 * sp.t.t1.3
SW.s.t1 =~ 1.0 * sw.s.t1.1 + .7 * sw.s.t1.2 + .7 * sw.s.t1.3

# factor covariance fixed @ 0.4 (medium effect size)
SP.o.t1 ~~ .4 * SP.t.t1
SP.o.t1 ~~ .4 * SW.s.t1
SP.t.t1 ~~ .4 * SW.s.t1

# factor variances fixed @ 0.4 (lowest number in H&S example)
SP.o.t1 ~~ 0.4 * SP.o.t1
SP.t.t1 ~~ 0.4 * SP.t.t1
SW.s.t1 ~~ 0.4 * SW.s.t1

# indicator residual variance fixed @ 0.40 (mean from sugus analysis)
sp.o.t1.1 ~~ .4 * sp.o.t1.1
sp.o.t1.2 ~~ .4 * sp.o.t1.2
sp.o.t1.3 ~~ .4 * sp.o.t1.3

sp.t.t1.1 ~~ .4 * sp.t.t1.1
sp.t.t1.2 ~~ .4 * sp.t.t1.2
sp.t.t1.3 ~~ .4 * sp.t.t1.3

sw.s.t1.1 ~~ .4 * sw.s.t1.1
sw.s.t1.2 ~~ .4 * sw.s.t1.2
sw.s.t1.3 ~~ .4 * sw.s.t1.3

'

# -------------------------------------------------------------------------- #
## Analysis model ----
# -------------------------------------------------------------------------- #

mod1.analyze <- '

# factorial structure
SP.o.t1 =~ sp.o.t1.1 + sp.o.t1.2 + sp.o.t1.3
SP.t.t1 =~ sp.t.t1.1 + sp.t.t1.2 + sp.t.t1.3
SW.s.t1 =~ sw.s.t1.1 + sw.s.t1.2 + sw.s.t1.3

'

# -------------------------------------------------------------------------- #
## Define missing pattern ----
# -------------------------------------------------------------------------- #

sim4.miss <- miss(pmMCAR = 0.2)   # 2% MCAR on all indicators (t1)

# -------------------------------------------------------------------------- #
## Define non-normal distributions for indicators ----
# -------------------------------------------------------------------------- #

# Predifine parameters of deviance from normality for 12 indicators
skew <- seq(-2, 2, length.out = 9)
kurt <- seq(-2, 2, length.out = 9)

# Bind 12 normal distributions with deviance parameters
dist <- bindDist(skewness = skew,
                 kurtosis = kurt)


# -------------------------------------------------------------------------- #
## Simulation ----
# -------------------------------------------------------------------------- #

sim4 <- sim(nRep = NULL,                     # number of replications for given sample size (see below), i.e. simulation repetitions
            model = mod1.analyze,            # analysis model
            generate = mod1.pop,             # population model
            n = rep(seq(900, 1000, 50), 100),  # n: target sample size
                                                # seq: varying sample sizes from A to B by increment C
                                                # rep: number of samples (= 1st argument)
            lavaanfun = "cfa",               # lavaan function
            std.lv = FALSE,                  # Standardization of latent variable not needed due to marker-variable method (1st loading fixed @ 1.00)
            multicore = FALSE,
            miss = sim4.miss,
            indDist = dist,
            estimator = "ml",
            maxit = 5000)

summaryFit(sim4)`
